client/instrumentation/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use parking_lot::Mutex;
+use tracing_core::{event::Event, Level, metadata::Metadata, span::{Attributes, Id, Record}, subscriber::Subscriber};
+
+use substrate_telemetry::{telemetry, SUBSTRATE_INFO};
+
+pub enum Receiver {
+	Telemetry,
+}
+
+#[derive(Debug, Default)]
+struct SpanDatum {
+	id: u64,
+	name: &'static str,
+	target: String,
+	line: u32,
+	start_time: u64,
+	overall_time: u64,
+}
+
+pub struct ProfilingSubscriber {
+	next_id: AtomicU64,
+	clock: quanta::Clock,
+	targets: Vec<(String, Level)>,
+	receiver: Receiver,
+	span_data: Mutex<Vec<SpanDatum>>,
+}
+
+impl ProfilingSubscriber {
+	pub fn new(receiver: Receiver, targets: &String) -> Self {
+		let targets: Vec<_> = targets.split(',').map(|s| parse_target(s)).collect();
+		ProfilingSubscriber {
+			next_id: AtomicU64::new(1),
+			clock: quanta::Clock::new(),
+			targets,
+			receiver,
+			span_data: Mutex::new(Vec::new()),
+		}
+	}
+}
+
+// Default to TRACE if no level given or unable to parse Level
+// We do not support a global `Level` currently
+fn parse_target(s: &str) -> (String, Level) {
+	match s.find("=") {
+		Some(i) => {
+			let a = s[0..i].to_string();
+			if s.len() > i {
+				let l = s[i + 1..s.len()].parse::<Level>().unwrap_or(Level::TRACE);
+				(a, l)
+			} else {
+				(a, Level::TRACE)
+			}
+		}
+		None => (s.to_string(), Level::TRACE)
+	}
+}
+
+impl Subscriber for ProfilingSubscriber {
+	fn enabled(&self, metadata: &Metadata<'_>) -> bool {
+		for x in &self.targets {
+			if metadata.target().starts_with(x.0.as_str()) && metadata.level() <= &x.1 {
```

Comment:
Try to use more meaningful names and maybe transform `x` from tuple to struct.

Reply:
I've swapped out `x` for `t` here, but seeing as it's a small private fn and concise 1 letter variable is easier to read - I think is best compromise. I did however follow your suggestion and updated some other variables in this module to have better names.